### PR TITLE
Boyer More Algorithm

### PR DESCRIPTION
**Pull request by DuckNrOne**
Hello,
I implemented the Boyer More Algorithm to find a pattern in a text.
I annouced it in #111.

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the comments
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

File: Algorithms/Search/Substring/BoyerMoore.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    /// Source: https://www.inf.hs-flensburg.de/lang/algorithmen/pattern/bmen.htm
+    /// Complexity:
+    ///     Time:   Preprocessing: O(a + m²)
+    ///             Comparison: O(n/m)
+    ///     Space:  O(m + a)
+    ///     where   m - pattern length
+    ///             n - text length
+    ///             a - alphabet length.
+    /// </summary>
+    public class BoyerMoore
```
Comment (siriak) [inline Algorithms/Search/Substring/BoyerMoore.cs:20]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277998)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Algorithms.Tests/Search/Substring/BoyerMoreTests.cs  1
- Algorithms/Search/Substring/BoyerMoore.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277998&bid=20433177)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms.Tests/Search/Substring/BoyerMoreTests.cs  1
- Algorithms/Search/Substring/BoyerMoore.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277998&bid=20433177)

File: Algorithms/Search/Substring/BoyerMoore.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        /// <summary>
+        /// Finds out the position of its rightmost occurrence in the pattern for each symbol of the alphabet,
+        /// or -1 if the symbol does not occur in the pattern.
+        /// </summary>
+        /// <param name="p">Search pattern.</param>
+        /// <param name="m">Length of the pattern.</param>
+        /// <returns>Array of the named postition for each symbol of the alphabet.</returns>
+        private static int[] BadCharacterRule(string p, int m)
+        {
+            // For each ASCII character
```
[the remark this answers is not in the document]
Comment (siriak) [reply Algorithms/Search/Substring/BoyerMoore.cs:85]: And also, char is 2 bytes, so not all chars will fit here, but I think it's fine to ignore it here for the sake of simplicity, just having a comment about that would be enough

**Review by siriak — APPROVED** [review]

File: Algorithms/Search/Substring/BoyerMoore.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (DuckNrOne) [reply Algorithms/Search/Substring/BoyerMoore.cs:85]: Thank you for the hint and the explanation :) I'll fix it asap

File: Algorithms/Search/Substring/BoyerMoore.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    /// <summary>
+    /// The idea:   You compare the pattern with the text from right to lef.
+    ///             If the text symbol that is compared with the rightmost pattern symbol
+    ///             does not occur in the pattern at all, then the pattern can be shifted
+    ///             by m positions behind this text symbol.
+    /// Source: https://www.inf.hs-flensburg.de/lang/algorithmen/pattern/bmen.htm
+    /// Complexity:
+    ///     Time:   Preprocessing: O(a + m²)
+    ///             Comparison: O(n/m)
+    ///     Space:  O(m + a)
```
[the remark this answers is not in the document]
Comment (DuckNrOne) [reply Algorithms/Search/Substring/BoyerMoore.cs:15]: There is already a link to my source that includes a explanation of the complexity

[the remark this answers is not in the document]
Comment (siriak) [reply Algorithms/Search/Substring/BoyerMoore.cs:14]: > The best case for the Boyer-Moore algorithm is attained if at each attempt the first compared text symbol does not occur in the pattern. Then the algorithm requires only O(n/m) comparisons.

It's best-case complexity, but usually, we write worst-case or both average and worst. Best-case is also valuable, just need to note that this is best-case complexity.

[the remark this answers is not in the document]
Comment (siriak) [reply Algorithms/Search/Substring/BoyerMoore.cs:14]: From Wikipedia
image
Could you write this in the comment?

[the remark this answers is not in the document]
Comment (DuckNrOne) [reply Algorithms/Search/Substring/BoyerMoore.cs:14]: Ofcourse :)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms.Tests/Search/Substring/BoyerMoreTests.cs  1
- Algorithms/Search/Substring/BoyerMoore.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277998&bid=20433177)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms.Tests/Search/Substring/BoyerMoreTests.cs  1
- Algorithms/Search/Substring/BoyerMoore.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277998&bid=20433177)

**Review by siriak — APPROVED** [review]
Thanks!

**Comment by DuckNrOne** [comment]
Sorry. I didn't mentioned this algorithm in README at all. Furthermore, the Boyer More Algorithm would be better off in Algorithm/Strings

**Pull request merged**